docs/data/material/customization/theme-components/theme-components.md

```diff
@@ -56,48 +56,114 @@ Each component is composed of several different parts.
+                border: `2px dashed ${blue[500]}`,
+              },
+            },
+            {
+              props: { variant: 'dashed', color: 'secondary' },
+              style: {
+                border: `4px dashed ${red[500]}`,
+              },
+            },
+          ],
+        },
       },
     },
   },
 });
 ```
 
-### Overrides based on props
+If you're using TypeScript, you'll need to specify your new variants/colors, using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation).
+
+<!-- Tested with packages/mui-material/test/typescript/augmentation/themeComponents.spec.ts -->
 
-You can pass a callback as a value in each slot of the component's `styleOverrides` to apply styles based on props.
+```tsx
+declare module '@mui/material/Button' {
+  interface ButtonPropsVariantOverrides {
+    dashed: true;
+  }
+}
+```
 
-The `ownerState` prop is a combination of public props that you pass to the component + internal state of the component.
+{{"demo": "GlobalThemeVariants.js"}}
+
+The variant `props` can also be defined as a callback.
+This is useful if you want to apply styles when using negation in the condition.
+In other words, applying a different style if a particular property doesn't have a specific value.
 
 ```js
-const finalTheme = createTheme({
+const theme = createTheme({
   components: {
-    MuiSlider: {
+    MuiButton: {
       styleOverrides: {
-        valueLabel: ({ ownerState, theme }) => ({
-          ...(ownerState.orientation === 'vertical' && {
-            backgroundColor: 'transparent',
-            color: theme.palette.grey[500],
-          }),
-        }),
+        root: {
+          variants: [
+            {
+              props: (props) =>
+                props.variant === 'dashed' && props.color !== 'secondary',
+              style: {
+                textTransform: 'none',
+                border: `2px dashed ${blue[500]}`,
+              },
+            },
+          ],
+        },
       },
     },
   },
 });
 ```
 
-{{"demo": "GlobalThemeOverrideCallback.js"}}
+#### Slot ownerState callback
+
+Using callback to access slot's `ownerState` has been deprecated. Consider migrating to `variants` using the codemod below:
+
+```bash
+npx @mui/codemod@next v6.0.0/theme-v6 <path/to/theme>
+```
+
+```diff
+ const finalTheme = createTheme({
+   components: {
+     MuiSlider: {
+       styleOverrides: {
+-        valueLabel: ({ ownerState, theme }) => ({
+-          ...(ownerState.orientation === 'vertical' && {
+-            backgroundColor: 'transparent',
+-            color: theme.palette.grey[500],
+-          }),
+-        }),
++        valueLabel: {
++          variants: [
++            props: { orientation: 'vertical' },
++            style: ({ theme }) => ({
++              backgroundColor: 'transparent',
++              color: theme.palette.grey[500],
++            }),
++          ],
++        }
+       },
+     },
+   },
+ });
+```
```

Comment:
Should this be added to https://mui.com/material-ui/migration/migrating-from-deprecated-apis? If we want to keep it here, I think we can just link to the Migration page and keep it short, removing the bash command and the example.

Reply:
Link to the migration works for me.